src/libUtils/UpgradeManager.cpp

```diff
@@ -15,148 +15,307 @@
 {
+    ((string*)userp)->append((char*)contents, size * nmemb);
+    return size * nmemb;
 }
 
-SWInfo::SWInfo(const SWInfo& src)
-    : m_major(src.m_major)
-    , m_minor(src.m_minor)
-    , m_fix(src.m_fix)
-    , m_upgradeDS(src.m_upgradeDS)
-    , m_commit(src.m_commit)
+string UpgradeManager::DownloadFile(const char* fileTail)
 {
-}
+    curl_global_init(CURL_GLOBAL_DEFAULT);
+    CURL* curl = curl_easy_init();
 
-SWInfo::~SWInfo(){};
+    if (!curl)
+    {
+        return nullptr;
+    }
 
-/// Implements the Serialize function inherited from Serializable.
-unsigned int SWInfo::Serialize(std::vector<unsigned char>& dst,
-                               unsigned int offset) const
-{
-    LOG_MARKER();
+    string curlRes;
+    curl_easy_setopt(curl, CURLOPT_URL, RELEASE_URL);
+    curl_easy_setopt(curl, CURLOPT_USERAGENT, USER_AGENT);
+    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
+    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &curlRes);
+    CURLcode res = curl_easy_perform(curl);
 
-    unsigned int size_remaining = dst.size() - offset;
+    if (res != CURLE_OK)
+        fprintf(stderr, "curl_easy_perform() failed: %s\n",
+                curl_easy_strerror(res));
 
-    if (size_remaining < SIZE)
+    curl_easy_cleanup(curl);
+    curl_global_cleanup();
+
+    int find = 0;
+    string cur;
+    vector<string> downloadFilePaths;
+    boost::char_separator<char> sep(",\"");
+    boost::tokenizer<boost::char_separator<char>> tokens(curlRes, sep);
+    for (boost::tokenizer<boost::char_separator<char>>::iterator tok_iter
+         = tokens.begin();
+         tok_iter != tokens.end(); ++tok_iter)
     {
-        dst.resize(SIZE + offset);
+        if (1 == find)
+        {
+            ++find;
+            continue;
+        }
+
+        if (2 == find)
+        {
+            find = 0;
+            downloadFilePaths.emplace_back(*tok_iter);
+            continue;
+        }
+
+        cur = *tok_iter;
+
+        if (cur == "browser_download_url")
+        {
+            find = 1;
+            continue;
+        }
     }
 
-    unsigned int curOffset = offset;
+    string downloadFilePath;
 
-    SetNumber<uint32_t>(dst, curOffset, m_major, sizeof(uint32_t));
-    curOffset += sizeof(uint32_t);
-    SetNumber<uint32_t>(dst, curOffset, m_minor, sizeof(uint32_t));
-    curOffset += sizeof(uint32_t);
-    SetNumber<uint32_t>(dst, curOffset, m_fix, sizeof(uint32_t));
-    curOffset += sizeof(uint32_t);
-    SetNumber<uint64_t>(dst, curOffset, m_upgradeDS, sizeof(uint64_t));
-    curOffset += sizeof(uint64_t);
-    SetNumber<uint32_t>(dst, curOffset, m_commit, sizeof(uint32_t));
-    curOffset += sizeof(uint32_t);
+    for (auto s : downloadFilePaths)
+    {
+        if (string::npos != s.rfind(fileTail))
+        {
+            downloadFilePath = s;
+            break;
+        }
+    }
 
-    return SIZE;
+    string cmd = downloadFilePath + " | wget -qi -";
+    return (system(cmd.c_str()) >= 0)
```

Comment:
Calling `system()` directly is quite dangerous and a better way is to use libraries to handle the downloading.